ksqldb-engine/src/main/java/io/confluent/ksql/internal/JmxDataPointsReporter.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.internal;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+
+public class JmxDataPointsReporter implements MetricsReporter {
+  private final Metrics metrics;
+  private final String group;
+  private final Map<MetricName, DataPointBasedGauge> gauges = new ConcurrentHashMap<>();
+  private final Duration staleThreshold;
+
+  public JmxDataPointsReporter(
+      final Metrics metrics,
+      final String group,
+      final Duration staleThreshold
+  ) {
+    this.metrics = Objects.requireNonNull(metrics, "metrics");
+    this.group = Objects.requireNonNull(group, "group");
+    this.staleThreshold = Objects.requireNonNull(staleThreshold, "staleThreshold");
+  }
+
+  @Override
+  public void report(final List<DataPoint> dataPoints) {
+    dataPoints.forEach(this::report);
+  }
+
+  private void report(final DataPoint dataPoint) {
+    final MetricName metricName
```

Comment:
would either the group of the datapoint contain the QueryID?

Reply:
yep - that's correct for metrics that are specific to a given query